### ISSUE TITLE
Huge memory consumption in finite TEBD time evolution

### ISSUE THREAD

**Issue by ErnestOng** [tenpy/tenpy#2]
The current TEBD code would end up consuming all of the available system memory after some time evolutions. The system used here is a spin chain with S=1.

The parameters for the simulation are:
```
chi = 70
L = 200
tmax = 9
dt = 0.02
```

The simulation grows and consumes all of the available RAM and pagefile at the point of the code:
```
    spectrum = np.zeros([int(tmax/dt),L-i0],dtype=complex)
    eng1 = tebd.Engine(psi, M, tebd_params)
    eng2 = tebd.Engine(psi_excited, M, tebd_params) #psi_excited is psi with 'Sp' applied at L//2
    for n in range(int(tmax / dt_measure + 0.5)):
        eng1.run()
        eng2.run()
        for i in range(i0+1,L,1):
            psitemp = psi_excited.copy()
            psitemp.apply_local_op(i, 'Sm', unitary=False)
            spectrum[n,i-i0-1] = psi.overlap(psitemp)[0]
```

**Comment by jhauschild** [comment]
Thanks for the report!
When you create the `tebd.Engine`, it should print you all parameters it's using (except if you set 'verbose' to <1). Can you tell me what it is using for  `trunc_par`?
In other words: did you check that the bond dimension is not growing beyond chi=70?

Apart from that, the last four lines of you code are a bit inefficient.
Try the following instead of the `for i in range` loop:
```python3
env = tenpy.networks.mps.MPSEnvironment(pis, psi_excited)
spectrum[n, :] = env.expectation_value('Sm')
```
This will calculate the expectation values you are aiming for at all sites more efficiently.

**Comment by ErnestOng** [comment]
Thank you for the reply!

The truncation parameter printed is
`
parameter 'trunc_params'={'chi_max' : 70, 'svd_min' : 1e-14, 'trunc_cut' : None} for TEBD
`

And looking through the outputs of the code, the max_chi remains at 70 for all time steps. There are a few timesteps that used the backup 'gesvd' lapack driver instead though. Not too sure if that matters.

The DMRG used to first obtain the ground state also has a chi of 70 except for the first and last 3 MPS but I believe that is correct due to OBC.

**Comment by jhauschild** [comment]
I can't replicate this problem on my machine. 
Can you provide a full minimal example where this happens? (Does it happen for the examples/b_intro_3_TEBD.py ?)
What OS are you on, what Python version are you using? Did you compile the cython parts?

You can try to explicitly tell the garbace collector to do its job, simply 
`import gc` and call `gc.collect()` in the loop over `n`, see [here](https://docs.python.org/3.5/library/gc.html).
You could also try to [trace back](https://docs.python.org/3/library/tracemalloc.html) where the leak is.

**Comment by ErnestOng** [comment]
Hi @jhauschild, so I have been trying to narrow down the problem using tracemalloc. After letting the code run for a day, python is using approximately 11GB ram but the outputs of malloc shows this:

/opt/intel/intelpython3/lib/python3.6/site-packages/tenpy/networks/mps.py:2202: size=321 MiB, count=8262742, average=41 B
/opt/intel/intelpython3/lib/python3.6/site-packages/tenpy/networks/mps.py:2200: size=139 MiB, count=6212335, average=23 B
/opt/intel/intelpython3/lib/python3.6/site-packages/tenpy/networks/mps.py:1835: size=81.3 MiB, count=3992101, average=21 B
/opt/intel/intelpython3/lib/python3.6/site-packages/tenpy/networks/mps.py:1812: size=74.1 MiB, count=3357853, average=23 B
/opt/intel/intelpython3/lib/python3.6/site-packages/tenpy/networks/mps.py:1836: size=55.4 MiB, count=2313062, average=25 B
/opt/intel/intelpython3/lib/python3.6/site-packages/tenpy/networks/mps.py:1813: size=48.0 MiB, count=2241270, average=22 B
/opt/intel/intelpython3/lib/python3.6/site-packages/tenpy/networks/mps.py:1841: size=47.2 MiB, count=280046, average=177 B
/opt/intel/intelpython3/lib/python3.6/site-packages/tenpy/algorithms/tebd.py:492: size=43.6 MiB, count=21512, average=2128 B
/opt/intel/intelpython3/lib/python3.6/site-packages/tenpy/algorithms/tebd.py:511: size=43.3 MiB, count=598, average=74.1 KiB
/opt/intel/intelpython3/lib/python3.6/site-packages/tenpy/networks/mps.py:170: size=26.5 MiB, count=1054212, average=26 B

`tracemalloc` is set to print the top 10 largest memory consumption. This is weird as the numbers do not really match up.

I am currently running the code on Ubuntu on a Windows Linux Subsystem environment. The Windows version is Windows 10 64bit. The python is intelpython 3.6.3 for Linux. Yes I did compile the cython parts to get the .so files using gcc 5.4.0.

I am now trying the same thing but with explicitly telling the garbage collector to clear up after every `n` iteration. Hopefully the consumption of ram issue can be at least temporarily solved.

**Comment by jhauschild** [comment]
To make sense of the line numbers: which tenpy version did you use exactly (i.e. what's the last git commit, or alternatively what's the output of `import tenpy; print(tenpy.version.full_version)`)?

Did you just take a single snapshot with `tracemalloc`? Maybe it would be more useful to compare the snapshots before and after ~10 calls to `eng.update()`, e.g.
```python3
snapshot1 = tracemalloc.take_snapshot()
for n in range(10):
        eng1.run()
        eng2.run()
snapshot2 = tracemalloc.take_snapshot()
top_stats = snapshot2.compare_to(snapshot1, 'lineno')
```
You shouldn't have to wait for a whole day to see the effect...

**Comment by ErnestOng** [comment]
@jhauschild, I have checked the version and it was 0.2.00. I think it was the TenPyLight that you sent to me. I have since git cloned the latest version and retried the same calculation. The same problem still happens and the output of `tracemalloc` is:

/home/ernest/tenpy/tenpy/networks/mps.py:2442: size=326 MiB (+326 MiB), count=8497876 (+8497876), average=40 B                                                                                                                                                                 /home/ernest/tenpy/tenpy/networks/mps.py:2439: size=143 MiB (+143 MiB), count=6403500 (+6403500), average=23 B                                                                                                                                                                 /home/ernest/tenpy/tenpy/networks/mps.py:1350: size=83.2 MiB (+83.1 MiB), count=4083346 (+4078181), average=21 B                                                                                                                                                               /home/ernest/tenpy/tenpy/networks/mps.py:1327: size=75.3 MiB (+75.3 MiB), count=3393349 (+3389325), average=23 B                                                                                                                                                               /home/ernest/tenpy/tenpy/networks/mps.py:1352: size=56.7 MiB (+56.6 MiB), count=2365686 (+2362502), average=25 B                                                                                                                                                               /home/ernest/tenpy/tenpy/networks/mps.py:1328: size=48.9 MiB (+48.8 MiB), count=2273781 (+2271379), average=23 B                                                                                                                                                               /home/ernest/tenpy/tenpy/algorithms/tebd.py:491: size=43.6 MiB (+43.6 MiB), count=21584 (+21584), average=2120 B                                                                                                                                                               /home/ernest/tenpy/tenpy/algorithms/tebd.py:510: size=43.3 MiB (+43.3 MiB), count=598 (+598), average=74.1 KiB                                                                                                                                                                 /home/ernest/tenpy/tenpy/networks/mps.py:171: size=27.1 MiB (+27.0 MiB), count=1078892 (+1076639), average=26 B                                                                                                                                                                /home/ernest/tenpy/tenpy/networks/mps.py:1357: size=47.2 MiB (+25.4 MiB), count=289356 (+286025), average=171 B 

Right now, the code is consuming 16GB of Ram. I have also tried running on a native linux environment using Ubuntu 16.04 and the same is happening where the amount of ram consumed is 13GB.

> You shouldn't have to wait for a whole day to see the effect...
The thing is the leak is very very slow and only after running over a day then you will notice the huge memory consumption.

I noticed one thing though, when I first git cloned the library, I forgot to compile the the np_conserved and charge python files. During that time, the memory consumption is much lesser though I did not run it for as long as a day to see if the memory was leaking. I have a hunch it is due to the low level c libraries that we are encountering this problem. 

If you would like, I could delete the .so files to run the code in the uncompiled form. I think it should pin point the problem to the cython files.

**Comment by jhauschild** [comment]
I could reproduce a similar problem when simply running the same DMRG code multiple times.
The Memory leak does *not* appear when you don't compile the Cython parts. Also, whether the memory leak appears seems to depend on the python/cython version used.
In particular, I had a memory leak when using intelpython (or anaconda), but not when using the default python3 shipped by Ubuntu. You could try using another python version on windows as well...

**Comment by jhauschild** [comment]
The following script for me produces a small memory leak for the current master branch when compiling the cython modules with intelpython or anaconda.

```python3
import numpy as np
import tenpy.linalg.np_conserved as npc
from tenpy.tools.process import memory_usage
import tenpy

tenpy.tools.optimization.set_level(1)

chinfo = npc.ChargeInfo()
leg = tenpy.linalg.charges.LegCharge(chinfo, [0, 10], [[]])#.bunch()[1]

A = npc.Array.from_func(np.random.random, [leg, leg.conj()], shape_kw='size')
A.iset_leg_labels(['vL', 'vR'])

def f():
    #  leg2 = leg.conj()
    U = A.copy()
    #  C /= npc.norm(C)
    C = npc.tensordot(A, U, axes=['vL', 'vR'])
    #  U, S, V = npc.svd(A, compute_uv=True)

N = 1000
for i in range(10):
    print("Mem:", memory_usage())
    for i in range(N):
        f()
print("Mem:", memory_usage())
```

Please confirm that this is the case for you as well.

I've tracked it down and noticed that at least on my machine the memory leak disappears when doing the changes of pull request #3. This should be at least a temporary solution to the problem - however as far as I understand Cythons np.ndarray syntax, the memory leak shouldn't appear in the first place...

**Comment by jhauschild** [comment]
Merged the fix #3, so I close this for now. If you still have a memory leak, please open a new issue...

**Comment by jhauschild** [comment]
I'm currently optimizing np_conserved a lot (although the code is not yet public),
and found the same issue again. I tracked it down to the following example code:

    import numpy as np
    import resource  # linux-only
    import memleak_ex

    def memory_usage():
        return resource.getrusage(resource.RUSAGE_SELF).ru_maxrss / 1024

    for shape in [(2, 2), (2, 0), (0, 2), (0, 0)]:
        print("shape", shape)
        x = np.zeros(shape, dtype=np.intp)
        N = 10000
        for i in range(5):
            print("Mem:", memory_usage())
            for i in range(N):
                memleak_ex.check_memleak(x)
        print("Mem:", memory_usage())

In combination with a cython file `memleak_ex.pyx` containing

    cimport numpy as cnp
    cimport cython

    cnp.import_array()

    def check_memleak(cnp.ndarray[cnp.intp_t, ndim=2] var):
        if var.shape[1] > 0:
            return False
        return True

The memory leak only appears when the second entry of the shape is 0. In TeNPy, this happend when qnumber was zero.
I guess I should open an issue in the cython repository...
